tests/5.0/declare_target/test_declare_target_nested_functions.F90

```diff
@@ -0,0 +1,60 @@
+!===--- test_declare_target_nested_functions.F90 -------------------------------------===//
+!
+! OpenMP API Version 5.0 Nov 2018
+!
+! Test of declare target on a host function that calls another host
+! function. According to 5.0 specification, the inner function, while
+! not explicitly declared on the target, should be treated as if it was.
+! The declared functions increment the passed-in value and return it. 
+! They are tested in a simple target region and the result is mapped out
+! and checked.
+!
+!//===----------------------------------------------------------------------===//
+
+#include "ompvv.F90"
+
+#define N 1024
+
+PROGRAM test_declare_target_nested_functions
+
+    USE iso_fortran_env
+    USE ompvv_lib
+    USE omp_lib
+    implicit none
+    INTEGER :: errors
+    errors = 0 
+
+    OMPVV_TEST_OFFLOADING
+
+    !$omp declare target to(outer_fn)
```

Comment:
This location is invalid. The OpenMP spec has:

> * If the directive does not have any clause, has a device_type clause, or has clauses that
contain a function name, it must appear in the specification part of a subroutine subprogram,
function subprogram or interface body.

The specification part of `program` ends with the assignment `errors = 1`.

Reply:
Actually, from that wording, it is not permitted in a `program` as this is neither a `subroutine` nor a `function` ...

Reply:
Thanks for you comment @tob2. What version of spec are you looking at? I see the following under declare target section for Version 5.0 (pg. 184, lines 7-10):

> • Any declare target directive with clauses must appear in a specification part of a subroutine subprogram, function subprogram, program or module.
> • Any declare target directive without clauses must appear in a specification part of a subroutine subprogram, function subprogram or interface body to which it applies.

By this I think it should work as long as it as moved to the specification part of the program, since it is a declare target with clauses. This is also consistent with 5.0 examples.

Reply:
> Thanks for you comment @tob2. What version of spec are you looking at? I see the following under declare target section for Version 5.0 (pg. 184, lines 7-10):

I did look at https://www.openmp.org/wp-content/uploads/openmp-TR9.pdf – however, this wording is already in TR8.

> > • Any declare target directive with clauses must appear in a specification part of a subroutine subprogram, function subprogram, program or module.

...

> By this I think it should work as long as it as moved to the specification part of the program, since it is a declare target with clauses. This is also consistent with 5.0 examples.

I concur that the 5.0.1 example `declare_target.2.f90` at https://www.openmp.org/wp-content/uploads/openmp-examples-5-0-1.pdf#page=190 shows a similar use. – And that the wording of 5.0 permits it.

In any case, it has to be in a specification part – whether it can be in the one of `program` or has to be in the one of `outer_fn`.

_Side remark: without further restrictions, OpenMP 5.0, TR8+TR9 permit code which is likely not intended to work (e.g. `declare target` for a use-associated variable). Hence, without a lengthy list of restrictions, it only makes sense to have a `declare target` for a procedure in the specification part of that procedure (or in the specification part an interface block for that procedure)._

_Hence, I think, it would be best to leave it out completely for this testcase. This should have no effect on the testcase, which relies already on auto-`declare target` of `inner_fn` such that the compiler can also auto-declare `outer_fn` without the directive._